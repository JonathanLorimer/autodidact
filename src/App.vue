<template>
	<div id="app">
		<header-nav></header-nav>
		<div class="spacer"></div>
		<router-view></router-view>
		<login id="login" v-if="loginSwitch"></login>
	</div>
</template>
<script>
export default {
	name: 'app',
	created(){
		this.$store.dispatch('getCookie', 'userCookie')
			.then( res => {
				if (res && !this.$store.state.users.currentUser) {
					this.$store.dispatch('loginUserById', res)
				}
			})
		this.$store.dispatch('getCookie', 'materialCookie')
			.then( res => {
				if (res) {
					this.$store.dispatch('setStateFromCookie', res)
				}
			})
	},
	data () {
		return {
		}
	},
	computed:{
		loginSwitch(){
			return this.$store.state.users.loginSwitch
		}
	},
	methods: {
		flipLoginSwitch(){
			this.$store.dispatch('flipLoginSwitch')
		}
	}
}
</script>
<style lang="scss">
#app {
	font-family: 'Avenir', Helvetica, Arial, sans-serif;
	-webkit-font-smoothing: antialiased;
	-moz-osx-font-smoothing: grayscale;
	text-align: center;
	color: #2c3e50;
	.spacer {
		height: 8rem;
	}
	position: relative;
}

#login {
	position: fixed;
	background-color: rgba(#2c3e50, .6);
	height: 100vh;
	width: 100vw;
	padding: 4rem;
	top: 0;
	left: 0;
}

</style>
