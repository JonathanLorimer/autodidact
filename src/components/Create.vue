	<template>
		<div class="container">
			<tree-struct
				class="root"
				:id="learningPath.id"
				:nodes="learningPath.nodes"
				:depth="0"
			>
			</tree-struct>
			<create-form 
				id="create-form" 
				v-if="createFormSwitch"
				@click.native="flipCreateSwitch"
				v-on:flipCreateSwitch="flipCreateSwitch"></create-form>
			<edit-form 
				id="edit-form" 
				v-if="editSwitch"
				@click.native="flipEditSwitch"
				v-on:flipEditSwitch="flipEditSwitch"></edit-form>
			<div class="button-container">
				<button class="add-material btn" @click="flipCreateSwitch">
					+ Add Learning Material
				</button>
				<button class="save-learning-path btn">
					Save Learning Path
				</button>
			</div>
		</div>
	</template>

	<script>
		export default {
			name: 'create',
			props: {
			},
			data () {
				return {
					createFormSwitch: false
				}
			},
			computed: {
				learningPath(){
					return this.$store.state.learningPath.learningPath
				},
				learningMaterials(){
					return this.$store.state.learningPath.learningMaterials
				},
				editSwitch(){
					return this.$store.state.learningPath.editSwitch
				}
			},
			mounted(){

			},
			methods: {
				flipCreateSwitch(){
					this.createFormSwitch = !this.createFormSwitch
				},
				flipEditSwitch(){
					this.$store.dispatch('flipEditSwitch')
				},
				traversePreReqs(obj, acc = []){
					acc.unshift(obj)
					if (!obj.preReq.name) return acc
					else return traversePreReqs(obj.preReq, acc)
				}
			}
		}
	</script>
	<style lang="scss" scoped>
		#create-form,
		#edit-form {
			position: fixed;
			background-color: rgba(#2c3e50, .6);
			height: 100vh;
			width: 100vw;
			padding: 4rem;
			top: 0;
			left: 0;
		}

		.root{
			background-color: #b7cec2;
			width: 80vw;
			min-height: 60vh;
			margin-top: 8rem;
			padding: 2rem 3rem 2rem 2rem;
			border-radius: .5rem;
			box-shadow: 0.5rem 1rem 3rem rgba(44, 62, 80, 0.1);
			display: flex;
			justify-content: center;
		}

		.btn {
			&:focus{
			outline: none;
			}
			border: none;
			box-shadow: 2px 5px 1rem rgba(44, 62, 80, 0.25);;
			border-radius: 50rem;
			font-size: 1.6rem;
			margin-top: 1rem;
			cursor: pointer;
			transition: all .2s;

			&:hover{
				transform: scale(1.2);
				box-shadow: 3px 7px 3rem rgba(44, 62, 80, 0.15);
			}

			&:active{
				transform: scale(1);
				box-shadow: 2px 5px 1rem rgba(44, 62, 80, 0.25);
			}
		}
		.button-container{
			display: flex;
			justify-content: space-evenly;
			width: 40vw;
			padding: 4rem;
			.add-material,
			.save-learning-path {
				background-color: #f37b56;
				color: #fff;
				height: 6rem;
				width: 24rem;
			}
		}

		.container {
			background-color: #e2f1e9;
			height: 120rem;
			display: flex;
			flex-direction: column;
			align-items: center;
			.learning_material {
				&__name-container {
					
				}
				&__name {

				}
				&__name-label {

				}
			}
		}

	</style>

