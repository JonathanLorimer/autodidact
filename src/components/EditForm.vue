<template>
	<div>
		<div class="container" @click.stop="doNothing">
			<form action="" class="learning-material" @submit.prevent="submitEdit">
				<div class="learning-material__name-container">
					<input
						type="text" 
						class="learning-material__name"
						v-model="learningMaterial.name"
						maxlength="100" 
						placeholder="Name"
						@click="clearErr">
					<label for="learning-material__name" class="learning-material__name-label"> Name </label>
				</div>
				<div class="learning-material__name-container">
					<input 
						list="material-type"
						class="learning-material__type"
						v-model="learningMaterial.type"
						maxlength="50" 
						placeholder="Type"
						@click="clearErr">
					<label for="learning-material__type" class="learning-material__type-label"> Type </label>
					<datalist id="material-type">
						<option value="Book"></option>
						<option value="Article"></option>
						<option value="Video"></option>
						<option value="Course"></option>
						<option value="Project"></option>
					</datalist>
				</div>
				<div class="learning-material__link-container">
					<input
						type="url" 
						class="learning-material__link"
						v-model="learningMaterial.link"
						maxlength="100" 
						placeholder="Link"
						@click="clearErr">
					<label for="learning-material__link" class="learning-material__link-label"> Link </label>
				</div>
				<div class="learning-material__pre-req-container">
					<input 
						class="learning-material__pre-req"
						:placeholder="learningMaterial.preReq.name || 'Pre-Requisite'"
						disabled>
					<div 
						class="event-catcher" 
						style="position:absolute; left:0; right:0; top:0; bottom:0;"
						@click="flipDropdownSwitch">
						<input type="checkbox" class="menu-checkbox" :checked="dropdownSwitch">
						<label for="menu-checkbox" class="menu-indicator"></label>
					</div>
					<div class="learning-material__pre-req-container--dropdown-container" v-show="dropdownSwitch">
						<ul class="learning-material__pre-req-container--dropdown">
							<li 
								class="learning-material__pre-req-container--item" 
								v-for="item in learningMaterials" 
								:key="item.id" 
								@click="addPreReq($event, item)"> {{ item.name }} </li>
							<li class="learning-material__pre-req-container--item" @click="addPreReq($event, false)">Remove</li>
						</ul>
					</div>
				</div>
				<div class="learning-material__button-container">
					<button class="learning-material__submit btn" type="submit"> Submit </button>
					<a class="learning-material__delete btn" @click.prevent="deleteMaterial"> Delete </a>
				</div>
				<div class="err">{{this.err}}</div>
			</form>
		</div>
	</div>
</template>

<script>
	export default {
		name: 'edit-form',
		props: {
		},
		data () {
			return {
				learningMaterial: {
					id: '',
					name: '',
					type: '',
					link: '',
					preReq: {},
				},
				dropdownSwitch: false,
				err: '',
			}
		},
		computed: {
			learningPath(){
				return this.$store.state.learningPath.learningPath
			},
			learningMaterials(){
				return Object.values(this.$store.state.learningPath.learningMaterials).filter(e => e.id !== this.learningMaterial.id)
			}
		},
		mounted(){
			this.learningMaterial.id = this.$store.state.learningPath.editing.id
			this.learningMaterial.name = this.$store.state.learningPath.editing.name
			this.learningMaterial.type = this.$store.state.learningPath.editing.type
			this.learningMaterial.link = this.$store.state.learningPath.editing.link
			this.learningMaterial.preReq = this.$store.state.learningPath.editing.preReq
		},
		methods: {
			clearErr(){
				this.err = ''
			},
			flipDropdownSwitch(){
				if(this.learningMaterials.length > 0) this.dropdownSwitch = !this.dropdownSwitch;
				else this.err = 'No learning materials to add as pre-requisite'
			},
			submitEdit(){
				// Trim white space from string entries
				this.learningMaterial.name = this.learningMaterial.name.trim()
				this.learningMaterial.type = this.learningMaterial.type.trim()
				this.learningMaterial.link = this.learningMaterial.link.trim()

				// Validation
				if (!this.learningMaterial.name) this.err = 'Please name your learning material'
				if (!this.learningMaterial.type) this.err = 'Please categorize your learning material with a type'

				// Save the Material and flip the form switch
				if(this.learningMaterial.name && this.learningMaterial.type) {
					this.$store.dispatch('editMaterial', this.learningMaterial)
					this.$emit('flipEditSwitch')
					this.learningMaterial = {
						name: '',
						type: '',
						link: '',
						preReq: {},
					}
				}
			},
			addPreReq(event, item){
				this.learningMaterial.preReq = item
				this.flipDropdownSwitch()
			},
			doNothing(){},
			deleteMaterial(){
				this.$store.dispatch('deleteMaterial', this.learningMaterial.id)
				this.$emit('flipEditSwitch')
			}
		}
	}
</script>
<style lang="scss" scoped>
	.btn {
		&:focus{
		outline: none;
		}
		border: none;
		box-shadow: none;
		border-radius: 50rem;
		font-size: 1.6rem;
		margin-top: 1rem;
		cursor: pointer;
		transition: all .2s;

		&:hover{
			transform: scale(1.2);
		}

		&:active{
			transform: scale(1)
		}
	}

	.container {
		position: absolute;
		top: 50%;
		left: 50%;
		transform: translate(-50%, -50%);
		height: 50rem;
		width: 46rem;
		border-radius: 3px;
		padding: 4rem 4rem 10rem;
		box-shadow: 1rem 1.5rem 2rem rgba(44, 62, 80, 0.3);
		background-color: #e2f1e9;
		font-size: 2rem;

		.learning-material {
			display: flex;
			flex-direction: column;
			justify-content: space-around;
			align-items: center;
			height: 100%;
			position: relative;
			
			.err {
				position: absolute;
				bottom: -6rem;
				width: 100%;
			}

			input:focus::-webkit-input-placeholder { color:transparent; }
			input:focus:-moz-placeholder { color:transparent; } /* FF 4-18 */
			input:focus::-moz-placeholder { color:transparent; } /* FF 19+ */
			input:focus:-ms-input-placeholder { color:transparent; }

			input::-webkit-input-placeholder { color:#2c3e50; font-size: 1.4rem; }
			input:-moz-placeholder { color:#2c3e50; font-size: 1.4rem; }
			input::-moz-placeholder { color:#2c3e50; font-size: 1.4rem; }
			input:-ms-input-placeholder { color:#2c3e50; font-size: 1.4rem; }
			select {
				-webkit-appearance: none;
				-moz-appearance: none;
				appearance: none;
			}

			&__name-container,
			&__type-container,
			&__link-container {
				position: relative;
			}

			&__name-label,
			&__type-label,
			&__link-label,
			&__pre-req-label {
				position: absolute;
				top: 50%;
				transform: translateY(-45%);
				left: 1rem;
				color: transparent;
				pointer-events: none;
				font-size: 1.5rem;
			}
			&__name:focus + .learning-material__name-label,
			&__type:focus + .learning-material__type-label,
			&__link:focus + .learning-material__link-label,
			&__pre-req:focus + .learning-material__pre-req-label, {
				position: absolute;
				top: -1rem;
				left: 0;
				color:#2c3e50;
				transition: all .4s cubic-bezier(.99,.15,.45,1.46)
			}

			&__name,
			&__name:active,
			&__name:focus,
			&__type,
			&__type:active,
			&__type:focus,
			&__link,
			&__link:active,
			&__link:focus,
			&__pre-req,
			&__pre-req:active,
			&__pre-req:focus {
				border: none;
				outline: none;
				width: 30rem;
				height: 3rem;
				padding-left: 1rem;
				font-size: 1.5rem;
			}

			&__submit {
				height: 4rem;
				width: 15rem;
			}

			&__delete {
				margin-left: 4rem;
			}

			&__button-container {
				display: flex;
				align-items: center
			}

			&__pre-req-container {
				position: relative;
				.event-catcher {
					position:absolute; 
					left:0; 
					right:0; 
					top:0; 
					bottom:0;
				}
				.menu-indicator {
					position: absolute;
					left: 27rem;
					top: 1.5rem;
					height: 2px;
					width: 1.8rem;
					background-color: black;
					&::before{
						content: '';
						position: absolute;
						top: .5rem;
						left: 0;
						height: 2px;
						width: 1.8rem;
						background-color: black;
						transition: all .3s ease;
					}
					&::after {
						content: '';
						position: absolute;
						top: -.5rem;
						left: 0;
						height: 2px;
						width: 1.8rem;
						background-color: black;
						transition: all .3s ease;
					}
				}
				.menu-checkbox {
					display: none;
				}
				.menu-checkbox:checked + .menu-indicator{
					background-color: transparent;
					&::before{
						transform: rotate(315deg);
						width: 1rem;
						top: .3rem;
						left: .8rem;
						transition: all .3 ease;
					}
					&::after {
						transform: rotate(-315deg);
						width: 1rem;
						top: -.3rem;
						left: .8rem;
						transition: all .3 ease;
					}
				}
				.event-catcher:hover .menu-checkbox:checked + .menu-indicator{
					&::before{
						top: .3rem;
					}
					&::after {
						top: -.3rem;
					}
				}
				.event-catcher:hover .menu-indicator {
					&:before {
						top: .7rem;
					}
					&:after {
						top: -.7rem;
					}
				}
				&--dropdown-container {
					position: absolute;
					background-color: #fff;
					left: 40rem;
					top: -26.3rem;
					height: 50rem;
					&::after {
						content: '';
						height: 10rem;
						width: 10rem;
						transform: rotate(45deg);
						position: absolute;
						background-color: #fff;
						top: 22.8rem;
						left: -5rem;
						z-index: -1;
					}
				}
				&--dropdown {
					max-height: 40rem;
					overflow-y: auto;
					width: 20rem;
					&::-webkit-scrollbar {
						background-color: #fff;
					}
					&::-webkit-scrollbar-thumb {
						background-color: #2c3e50;
					}
				}
				&--item{
					width: 100%;
					list-style: none;
					padding: .5rem 4rem;
					cursor: pointer;
				}
				&--item:hover{
					background-color:#2c3e50;
					color: #fff;
				}

			}
		}
	}

</style>