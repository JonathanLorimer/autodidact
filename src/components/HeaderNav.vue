	<template>
		<div class="navigation">
			<nav class="navigation__nav">
				<div class="navigation__logo"><router-link class="link" :to="'/'">Home</router-link></div>
				<ul class="navigation__list">
					<li class="navigation__item"><router-link class="link" :to="'create'">Create</router-link></li>
					<li class="navigation__item"><router-link class="link" :to="'view-proj'">View Projects</router-link></li>
					<li class="navigation__item" v-if="!currentUser"><a class="link login" @click="flipLoginSwitch"> Login/Register </a></li>
					<li class="navigation__item" v-else><a class="link user"> {{ currentUser }} </a><a class="link logout" @click="logout"> Logout </a> </li>
				</ul>
			</nav>
		</div>
	</template>

	<script>
		export default {
			name: 'header-nav',
			props: {
			},
			data () {
				return {
				}
			},
			computed: {
				currentUser(){
					return this.$store.state.users.currentUser
				}
			},
			methods: {
				flipLoginSwitch(){
					this.$store.dispatch('flipLoginSwitch')
				},
				logout(){
					this.$store.dispatch('logout')
					this.$store.dispatch('delCookie', 'userCookie')
				}
			}
		}
	</script>
	<style lang="scss" scoped>
		li {
			list-style: none;
		}

		.link,
		.link:active,
		.link:visited {
			text-decoration: none;
			color: #fff;
			cursor: pointer;
		}
		
		.navigation {
			background-color: #3c4859;
			height: 8rem;
			position: fixed;
			top: 0;
			left: 0;
			width: 100%;
			height: 8rem;
			z-index: 1;
			padding-left: 3rem;

			&__nav  {
				height: 100%;
				display: flex;
				justify-content: space-between;
				align-items: center;
			}

			&__logo {
				font-size: 2.4rem;
			}

			&__list {
				width: 50%;
				display: flex;
				justify-content: space-evenly;
			}

			&__item {
				font-size: 1.6rem;
				color: #fff;
			}
		}
	</style>

