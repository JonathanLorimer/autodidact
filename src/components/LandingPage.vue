	<template>
		<div class="landing-page">
			<section class="hero-image">
				<div class="header">
					<h1 class="header__text">Autodidact</h1>
					<p class="header__content">A tool for the self learner to connect and organize their materials</p>
				</div>
			</section>
		</div>
	</template>

	<script>
		export default {
			name: 'landing-page',
			props: {
			},
			data () {
				return  {

				}
			}
		}
	</script>
	<style lang="scss" scoped>
		li {
			list-style: none;
		}
		.landing-page {
			background-color: #e2f1e9;
			height: 200rem;
		}
		.hero-image {
			display: flex;
			align-content: center;
			justify-content: center;
			background-image: 
				linear-gradient(
					rgba(#f37b56, 0.8),
					rgba(#f37b56, 0.8)
				),
				url("../assets/book_background.png");
			background-size: cover;
			-webkit-clip-path: polygon(0 0, 100% 0, 100% 75vh, 0 100%);
			clip-path: polygon(0 0, 100% 0, 100% 75vh, 0 100%);
			height: 90vh;
			.header {
				display: flex;
				flex-direction: column;
				justify-content: center;
				color: #fff;
				&__text {
					font-size: 10rem;
					margin-bottom: 2rem;
				}
				&__content{
					font-size: 2.5rem;
					margin-bottom: 12vh;
				}
			}
		}
	</style>

