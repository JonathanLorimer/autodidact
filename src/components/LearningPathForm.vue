<template>
	<div>
		<div class="container" @click.stop="doNothing">
			<transition name="fade" mode="out-in">
				<form action="" class="learning-path fade" @submit.prevent="submitLearningPath" v-if="!success" :key="'form'">
				<!-- <form action="" class="learning-path fade" @submit.prevent="submitLearningPath" :key="'form'"> -->
					<div class="learning-path__name-container">
						<input
							type="text" 
							class="learning-path__name"
							v-model="learningPath.name"
							maxlength="100" 
							placeholder="Name"
							@click="clearErr">
						<label for="learning-path__name" class="learning-path__name-label"> Name </label>
					</div>
					<button class="learning-path__submit btn" type="submit" > Submit </button>
					<div class="err">{{this.err}}</div>
				</form>
				<div class="success-message fade" v-else :key="'success-message'">
				<!-- <div class="success-message fade" :key="'success-message'"> -->
					<div class="success-message__message">
						<img src="../assets/approved_check_checkbox_checkmark_confirm_success_yes_icon_128.svg" alt="" class="success-message__image">
						Learning Path successfully saved
					</div>
				</div>
			</transition>
		</div>
	</div>
</template>

<script>
	import uuid from 'uuid/v4'
	export default {
		name: 'learning-path-form',
		props: {
		},
		data () {
			return {
				learningPath: {
					name: '',
				},
				err: '',
				success: false,
			}
		},
		computed: {
		},
		mounted(){
		},
		methods: {
			clearErr(){
				this.err = ''
			},
			submitLearningPath(){
				// Trim white space from string entries
				this.learningPath.name = this.learningPath.name.trim()


				// Save the PathlearningPath and flip the form switch
				if (this.$store.state.learningPath.currentPathId){
					this.$store.dispatch('updateLearningPath')
					.then(res => {
						if (res.status.success){
							this.success = true
							setTimeout(() => {
								this.$emit('flipLPSwitch')
								this.learningPath = {
									name: '',
								}
								this.success = false
							}, 2500)
						}
					})
				}
				else if(this.learningPath.name) {
					let userId = this.$store.state.users.currentUserId
					let name = this.learningPath.name
					this.$store.dispatch('saveLearningPath', { userId, name })
					.then(res => {
						if (res.status.success){
							this.success = true
							setTimeout(() => {
								this.$emit('flipLPSwitch')
								this.learningPath = {
									name: '',
								}
								this.success = false
							}, 2500)
						}
					})
				}
			},
			doNothing(){}
		}
	}
</script>
<style lang="scss" scoped>
	// transition group
	.fade-enter-active {
	  transition: all .3s ease;
	}
	.fade-leave-active {
	  transition: all .8s cubic-bezier(1.0, 0.5, 0.8, 1.0);
	}
	.fade-enter, .fade-leave-to
	/* .slide-fade-leave-active below version 2.1.8 */ {
	  transform: translateX(10px);
	  opacity: 0;
	}
	.transition-wrapper {
		height: 100%;
		width: 100%;
	}

	.btn {
		&:focus{
		outline: none;
		}
		border: none;
		box-shadow: none;
		border-radius: 50rem;
		font-size: 1.6rem;
		margin-top: 1rem;
		cursor: pointer;
		transition: all .2s;

		&:hover{
			transform: scale(1.2);
		}

		&:active{
			transform: scale(1)
		}
	}

	.container {
		position: absolute;
		top: 50%;
		left: 50%;
		transform: translate(-50%, -50%);
		height: 30rem;
		width: 46rem;
		border-radius: 3px;
		padding: 4rem 4rem 10rem;
		box-shadow: 1rem 1.5rem 2rem rgba(44, 62, 80, 0.3);
		background-color: #e2f1e9;
		font-size: 2rem;

		.learning-path {
			display: flex;
			flex-direction: column;
			justify-content: space-around;
			align-items: center;
			height: 100%;
			position: relative;
			
			.err {
				position: absolute;
				bottom: -6rem;
				width: 100%;
			}

			input:focus::-webkit-input-placeholder { color:transparent; }
			input:focus:-moz-placeholder { color:transparent; } /* FF 4-18 */
			input:focus::-moz-placeholder { color:transparent; } /* FF 19+ */
			input:focus:-ms-input-placeholder { color:transparent; }

			input::-webkit-input-placeholder { color:#2c3e50; font-size: 1.4rem; }
			input:-moz-placeholder { color:#2c3e50; font-size: 1.4rem; }
			input::-moz-placeholder { color:#2c3e50; font-size: 1.4rem; }
			input:-ms-input-placeholder { color:#2c3e50; font-size: 1.4rem; }
			select {
				-webkit-appearance: none;
				-moz-appearance: none;
				appearance: none;
			}

			&__name-container {
				position: relative;
			}

			&__name-label {
				position: absolute;
				top: 50%;
				transform: translateY(-45%);
				left: 1rem;
				color: transparent;
				pointer-events: none;
				font-size: 1.5rem;
			}
			&__name:focus + .learning-path__name-label {
				position: absolute;
				top: -1rem;
				left: 0;
				color:#2c3e50;
				transition: all .4s cubic-bezier(.99,.15,.45,1.46)
			}

			&__name,
			&__name:active,
			&__name:focus {
				border: none;
				outline: none;
				width: 30rem;
				height: 3rem;
				padding-left: 1rem;
				font-size: 1.5rem;
			}

			&__submit {
				height: 4rem;
				width: 15rem;
			}
		}
		.success-message {
			height: 100%;
			width: 100%;
			display: flex;
			&__message{
				display: flex;
				align-items: center;
				margin-top: 4rem;
			}
			&__image {
				height: 4.8rem;
				width: 4.8rem;
				margin-right: 2rem;
			}
		}
	}
</style>