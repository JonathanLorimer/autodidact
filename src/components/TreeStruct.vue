<template>
  <div class="tree-struct">
	<div class="tree-struct__container" v-if="learningMaterials[id]">
		<tr class="tree-struct__row" :style="indent">
			<td class="tree-struct__block tree-struct__name"><a :href="learningMaterials[id].link" target="_blank" class="tree-struct__link">{{ learningMaterials[id].name }}</a></td>
			<td class="tree-struct__block tree-struct__type">{{ learningMaterials[id].type }}</td>
		</tr>
		<div class="tree-struct__edit"><a class="tree-struct__edit--link" @click="flipEditSwitch"><i class="far fa-edit"></i></a></div>
	</div>
    <tree-struct
      v-for="node in nodes"
	  :key ="node.id"
      :nodes="node.nodes" 
      :id="node.id"
	  :depth="depth + 1"
    >
    </tree-struct>
  </div>
</template>
<script>
  export default { 
    props: [ 'id', 'nodes', 'depth' ],
	name: 'tree-struct',
	data(){
		return{}
	},
	computed: {
		learningMaterials(){
			return this.$store.state.learningPath.learningMaterials
		},
		indent() {
        	return { marginLeft: `${(this.depth - 1) * 6}rem` }
      	}
	},
	methods: {
		flipEditSwitch(){
			this.$store.dispatch('flipEditSwitch', this.id)
		}
	}
  }
</script>
<style lang="scss" scoped>
	.tree-struct {
		&__container {
			display: flex;
		}
		&__container:hover .tree-struct__edit{
			display: flex;
		}
		&__edit{
			display: none;	
			background-color: #fff;
			height: 3.6rem;
			border-radius: 3px;
			width: 3.6rem;
			margin-left: .5rem;
			justify-content: center;
			align-items: center;
			&--link{
				height: 2rem;
				margin-bottom: .4rem;
				margin-left: .4rem;
				cursor: pointer;
			}
			.fa-edit,
			.fa-edit:link,
			.fa-edit:active,
			.fa-edit:visited {
				outline: none;
				color: #2c3e50;
				font-size: 2rem;
			}
		}
		&__row {
			// background-color: #e8edeb;
			background-color: #fff;
			height: 3rem;
			border-radius: 3px;
			display: flex;
			width: auto;
			align-items: center;
			justify-content: space-between;
			margin-bottom: 1rem;
			padding: 1.8rem 0;
			// TODO: figure out how to animate hovering over row
			// transition: all .2s;
			// &:hover {
			// 	outline: 3px solid #2c3e50;
			// }
		}

		&__block {
			padding: 0 1.5rem;
			font-size: 2rem;
		}
		&__link:link,
		&__link:active,
		&__link:visited {
			text-decoration: none;
			color:#2c3e50;
			font-weight: 900;
		}
	}
</style>
