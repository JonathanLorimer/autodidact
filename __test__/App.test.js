import Vue from 'vue'
import Vuex from 'vuex'
import VueRouter from 'vue-router'
import { mount, shallowMount, createLocalVue } from '@vue/test-utils'
import App from '../src/App'
import LandingPage from '../src/components/LandingPage.vue'
import HeaderNav from '../src/components/HeaderNav.vue'
import Login from '../src/components/Login.vue'
import Create from '../src/components/Create.vue'
import CreateForm from '../src/components/CreateForm.vue'
import EditForm from '../src/components/EditForm.vue'
import LearningPathForm from '../src/components/LearningPathForm.vue'
import ViewProj from '../src/components/ViewProj.vue'
import TreeStruct from '../src/components/TreeStruct.vue'

// cmp.component('login', Login)
// cmp.component('create', Create)
// cmp.component('create-form', CreateForm)
// cmp.component('edit-form', EditForm)
// cmp.component('learning-path-form', LearningPathForm)
// cmp.component('view-proj', ViewProj)
// cmp.component('tree-struct', TreeStruct)

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(VueRouter)

describe('Testing App.vue, LandingPage.vue, HeaderNav.vue', () => {
	let cmp, vm, actions, state, store, router
	beforeEach(() => {
		cmp = localVue.extend(App) // Create a copy of the original component
		cmp.component('landing-page', LandingPage) //Register components
		cmp.component('header-nav', HeaderNav)

		state = {
			users: {
				currentUser: false,
				loginSwitch: false
			}
		}
	
		actions = {
			flipLoginSwitch: jest.fn(),
			logout: jest.fn(),
			delCookie: jest.fn(),
			getCookie: jest.fn(),
			clearLocalStorage: jest.fn()
		}
	
		store = new Vuex.Store({
			state,
			actions
		})
		
		router = new VueRouter()

		vm = new cmp({
			store,
			router,
		}).$mount() // Instances and mounts the component
	})
	
	it('Landing Page should mount and render correctly (snapshot test)', () => {
		const wrapper = mount(LandingPage)
		expect(wrapper.html()).toMatchSnapshot()
	})

	it('Header Nav: Login button should call flipLoginSwitch ', () => {
		const wrapper = shallowMount(HeaderNav, { store, router, localVue })
		wrapper.find('.link.login').trigger('click')
		expect(actions.flipLoginSwitch).toBeCalled()
	})

	it('Header Nav: Logout button should call logout and delCookie ', () => {
		const wrapper = shallowMount(HeaderNav, { store, router, localVue })
		store.state.users.currentUser = true

		wrapper.find('.link.logout').trigger('click')
		expect(actions.logout).toBeCalled()
		expect(actions.delCookie.mock.calls).toHaveLength(1)
    	expect(actions.delCookie.mock.calls[0][1]).toEqual('userCookie')
	})

	it('Header Nav: should mount and render correctly (snapshot test)', () => {
		const wrapper = shallowMount(HeaderNav, { store, router, localVue })
		expect(wrapper.html()).toMatchSnapshot()
	})

	it('Header Nav: should mount and render correctly when user is logged in (snapshot test)', () => {
		const wrapper = shallowMount(HeaderNav, { store, router, localVue })
		store.state.users.currentUser = "testUser@hotmail.com"

		expect(wrapper.find('.link.user').text()).toBe('testUser@hotmail.com')
		expect(wrapper.html()).toMatchSnapshot()
	})

})