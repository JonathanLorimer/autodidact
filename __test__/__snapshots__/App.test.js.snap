// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing App.vue, LandingPage.vue, HeaderNav.vue Header Nav: should mount and render correctly (snapshot test) 1`] = `
<div class="navigation">
  <nav class="navigation__nav">
    <div class="navigation__logo">
      <vuecomponent-stub to="/" class="link"></vuecomponent-stub>
    </div>
    <ul class="navigation__list">
      <li class="navigation__item">
        <vuecomponent-stub to="create" class="link"></vuecomponent-stub>
      </li>
      <li class="navigation__item">
        <vuecomponent-stub to="view-proj" class="link"></vuecomponent-stub>
      </li>
      <li class="navigation__item">
        <a class="link login"> Login/Register </a>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`Testing App.vue, LandingPage.vue, HeaderNav.vue Header Nav: should mount and render correctly when user is logged in (snapshot test) 1`] = `
<div class="navigation">
  <nav class="navigation__nav">
    <div class="navigation__logo">
      <vuecomponent-stub to="/" class="link"></vuecomponent-stub>
    </div>
    <ul class="navigation__list">
      <li class="navigation__item">
        <vuecomponent-stub to="create" class="link"></vuecomponent-stub>
      </li>
      <li class="navigation__item">
        <vuecomponent-stub to="view-proj" class="link"></vuecomponent-stub>
      </li>
      <li class="navigation__item">
        <a class="link user"> testUser@hotmail.com </a>
        <a class="link logout"> Logout </a>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`Testing App.vue, LandingPage.vue, HeaderNav.vue Landing Page should mount and render correctly (snapshot test) 1`] = `
<div class="landing-page">
  <section class="hero-image">
    <div class="header">
      <h1 class="header__text">Autodidact</h1>
      <p class="header__content">A tool for the self learner to connect and organize their materials</p>
    </div>
  </section>
</div>
`;
